cpp/src/arrow/c/helpers.h

```diff
@@ -18,10 +18,20 @@
 #pragma once
 
 #include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
 
 #include "arrow/c/abi.h"
 
+#define CHECK_OR_DIE(condition, msg)                          \
+  do {                                                        \
+    if (!(condition)) {                                       \
+      fprintf(stderr, "%s:%d:: %s", __FILE__, __LINE__, msg); \
```

Comment:
Let's allow `msg` to be complex.
```suggestion
      fprintf(stderr, "%s:%d:: %s", __FILE__, __LINE__, (msg)); \
```